### ISSUE TITLE
[WebDAV] Can not replace http `Referer`?

### ISSUE THREAD

**Issue by nilaoda** [rclone/rclone#6040]
Rclone version: v1.57.0

---

my conf file
```ini
[AList]
type = webdav
url = http://192.168.2.2:5244/dav
vendor = other
user = *
pass = *
headers = "Referer","https://www.mycustom.com/drive","XXX","YYY"
```

log file
```log
2022/03/17 02:41:23 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2022/03/17 02:41:23 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2022/03/17 02:41:23 DEBUG : HTTP REQUEST (req 0xc00014f400)
2022/03/17 02:41:23 DEBUG : GET /** HTTP/1.1
Host: **
User-Agent: rclone/v1.57.0
Depth: 0
Range: bytes=341114880-682229759
Referer: http://192.168.2.2:5244/dav/%E8%A5%BF%E9%81%8A%E8%A8%98/MUX_1.mkv
Xxx: YYY
```

It doesn't seem to use the Referer in the conf file. And the Header's key is forced to be capitalized.

How can i custom the Referer?

**Comment by nilaoda** [comment]
Related code:

https://github.com/rclone/rclone/blob/807f1cedaac3de6c527acbae2c7e4c40cab35fbd/backend/webdav/webdav.go#L449-L458



Referer is forced to set after setting custom headers?

**Comment by animosity22** [comment]
@ncw - is this something to keep open / enhancement / close out?

**Comment by nilaoda** [comment]
> It would be easy enough to not set the `Referer` if there was one set already - do you want to try adding that @nilaoda ?

I want add specified referer to pass server referer policy

**Comment by ncw** [comment]
@nilaoda give this a go

[v1.59.0-beta.6085.82e24f521.fix-6040-webdav](https://beta.rclone.org/branch/fix-6040-webdav/v1.59.0-beta.6085.82e24f521.fix-6040-webdav/) on branch [fix-6040-webdav](https://github.com/rclone/rclone/tree/fix-6040-webdav) (uploaded in 15-30 mins)

It won't override Referer if you set it.


> And the Header's key is forced to be capitalized

That is go's HTTP library doing that. HTTP headers are case insensitive so it shouldn't matter.

**Comment by nilaoda** [comment]
> @nilaoda give this a go
> 
> [v1.59.0-beta.6085.82e24f521.fix-6040-webdav](https://beta.rclone.org/branch/fix-6040-webdav/v1.59.0-beta.6085.82e24f521.fix-6040-webdav/) on branch [fix-6040-webdav](https://github.com/rclone/rclone/tree/fix-6040-webdav) (uploaded in 15-30 mins)
> 
> It won't override Referer if you set it.
> 
> > And the Header's key is forced to be capitalized
> 
> That is go's HTTP library doing that. HTTP headers are case insensitive so it shouldn't matter.

Thx! ^_^

**Comment by ncw** [comment]
@nilaoda did it work properly? If so I'll merge it - thanks!

**Comment by nilaoda** [comment]
@ncw Still confused,

###   `User-Agent` not working?

conf:
```config
headers = "Referer","https://www.aliyundrive.com/drive","User-Agent","Mozilla/5.0 (Windows NT 6.1; Trident/7.0; rv:11.0) like Gecko"
```
log:
```
2022/04/26 17:50:35 DEBUG : PROPFIND /dav/2019.mkv HTTP/1.1
Host: 192.168.2.2:5244
User-Agent: rclone/v1.59.0-beta.6085.82e24f521.fix-6040-webdav
Authorization: XXXX
Depth: 1
Referer: https://www.aliyundrive.com/drive
Accept-Encoding: gzip
```

### Missing headers after Redirect ?
```
2022/04/26 17:50:35 DEBUG : GET /dav/2019.mkv HTTP/1.1
Host: 192.168.2.2:5244
User-Agent: rclone/v1.59.0-beta.6085.82e24f521.fix-6040-webdav
Authorization: XXXX
Depth: 0
Range: bytes=0-371982335
Referer: https://www.aliyundrive.com/drive


2022/04/26 17:50:35 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2022/04/26 17:50:36 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2022/04/26 17:50:36 DEBUG : HTTP RESPONSE (req 0xc0008fda00)
2022/04/26 17:50:36 DEBUG : HTTP/1.1 302 Found
Content-Length: 566
Content-Type: text/html; charset=utf-8
Date: Tue, 26 Apr 2022 09:50:36 GMT
Etag: "16e9663435815b802c574532"
Location: https://bj29.cn-beijing.data.alicloudccp.com/****


2022/04/26 17:50:36 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2022/04/26 17:50:36 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2022/04/26 17:50:36 DEBUG : HTTP REQUEST (req 0xc0008fdc00)
2022/04/26 17:50:36 DEBUG : GET /**** HTTP/1.1
Host: bj29.cn-beijing.data.alicloudccp.com
User-Agent: rclone/v1.59.0-beta.6085.82e24f521.fix-6040-webdav
Depth: 0
Range: bytes=371982336-743916849
Referer: http://192.168.2.2:5244/dav/2019.mkv
```

**Comment by ncw** [comment]
> 2022/04/26 17:50:36 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
> 2022/04/26 17:50:36 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> 2022/04/26 17:50:36 DEBUG : HTTP REQUEST (req 0xc0008fdc00)
> 2022/04/26 17:50:36 DEBUG : GET /**** HTTP/1.1
> Host: bj29.cn-beijing.data.alicloudccp.com
> User-Agent: rclone/v1.59.0-beta.6085.82e24f521.fix-6040-webdav
> Depth: 0
> Range: bytes=371982336-743916849
> Referer: http://192.168.2.2:5244/dav/2019.mkv

I think that is the go standard library doing that - see https://github.com/golang/go/issues/44160 so I don't think I can work around that at the moment so easily.

I've merged the fix above to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.59

**Comment by nilaoda** [comment]
> > 2022/04/26 17:50:36 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
> > 2022/04/26 17:50:36 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> > 2022/04/26 17:50:36 DEBUG : HTTP REQUEST (req 0xc0008fdc00)
> > 2022/04/26 17:50:36 DEBUG : GET /**** HTTP/1.1
> > Host: bj29.cn-beijing.data.alicloudccp.com
> > User-Agent: rclone/v1.59.0-beta.6085.82e24f521.fix-6040-webdav
> > Depth: 0
> > Range: bytes=371982336-743916849
> > Referer: http://192.168.2.2:5244/dav/2019.mkv
> 
> I think that is the go standard library doing that - see [golang/go#44160](https://github.com/golang/go/issues/44160) so I don't think I can work around that at the moment so easily.
> 
> I've merged the fix above to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.59

Emm... Okay, I will wait for that. Maybe it is necessary to disable the automatic redirection and manual handle `302`, `301` response code.

Thanks for your help. 👍

**Comment by flymin** [comment]
I add this line to the config:
```
headers = "Referer","https://www.aliyundrive.com/drive","User-Agent","Mozilla/5.0 (Windows NT 6.1; Trident/7.0; rv:11.0) like Gecko"
```
However, the `rclone copy` still report 403, as follows:
```
> rclone -vv copy alist-aliyundrive:alist-backup/"alist_backup_1_29_2023, 11_49_39 AM.json" .

2023/02/12 13:32:15 DEBUG : rclone: Version "v1.61.1" starting with parameters ["rclone" "-vv" "copy" "alist-aliyundrive:alist-backup/alist_backup_1_29_2023, 11_49_39 AM.json" "."]
2023/02/12 13:32:15 DEBUG : Creating backend with remote "alist-aliyundrive:alist-backup/alist_backup_1_29_2023, 11_49_39 AM.json"
2023/02/12 13:32:15 DEBUG : Using config file from "/Users/gry/.config/rclone/rclone.conf"
2023/02/12 13:32:15 DEBUG : found headers: Referer,https://www.aliyundrive.com/drive,User-Agent,Mozilla/5.0 (Windows NT 6.1; Trident/7.0; rv:11.0) like Gecko
2023/02/12 13:32:15 DEBUG : fs cache: adding new entry for parent of "alist-aliyundrive:alist-backup/alist_backup_1_29_2023, 11_49_39 AM.json", "alist-aliyundrive:alist-backup"
2023/02/12 13:32:15 DEBUG : Creating backend with remote "."
2023/02/12 13:32:15 DEBUG : fs cache: renaming cache item "." to be canonical "/Users/gry"
2023/02/12 13:32:15 DEBUG : alist_backup_1_29_2023, 11_49_39 AM.json: Need to transfer - File not found at Destination
2023/02/12 13:32:16 ERROR : alist_backup_1_29_2023, 11_49_39 AM.json: Failed to copy: failed to open source object: 403 Forbidden
2023/02/12 13:32:16 ERROR : Attempt 1/3 failed with 1 errors and: failed to open source object: 403 Forbidden
2023/02/12 13:32:16 DEBUG : alist_backup_1_29_2023, 11_49_39 AM.json: Need to transfer - File not found at Destination
2023/02/12 13:32:16 ERROR : alist_backup_1_29_2023, 11_49_39 AM.json: Failed to copy: failed to open source object: 403 Forbidden
2023/02/12 13:32:16 ERROR : Attempt 2/3 failed with 1 errors and: failed to open source object: 403 Forbidden
2023/02/12 13:32:16 DEBUG : alist_backup_1_29_2023, 11_49_39 AM.json: Need to transfer - File not found at Destination
2023/02/12 13:32:16 ERROR : alist_backup_1_29_2023, 11_49_39 AM.json: Failed to copy: failed to open source object: 403 Forbidden
2023/02/12 13:32:16 ERROR : Attempt 3/3 failed with 1 errors and: failed to open source object: 403 Forbidden
2023/02/12 13:32:16 INFO  :
Transferred:   	          0 B / 0 B, -, 0 B/s, ETA -
Errors:                 1 (retrying may help)
Elapsed time:         1.4s

2023/02/12 13:32:16 DEBUG : 8 go routines active
2023/02/12 13:32:16 Failed to copy: failed to open source object: 403 Forbidden
```

If we execute with `--header="Referer: https://www.aliyundrive.com/drive"`, it works:
```
> rclone -vv copy alist-aliyundrive:alist-backup/"alist_backup_1_29_2023, 11_49_39 AM.json" . --header="Referer: https://www.aliyundrive.com/drive"

2023/02/12 13:34:24 DEBUG : rclone: Version "v1.61.1" starting with parameters ["rclone" "-vv" "copy" "alist-aliyundrive:alist-backup/alist_backup_1_29_2023, 11_49_39 AM.json" "." "--header=Referer: https://www.aliyundrive.com/drive"]
2023/02/12 13:34:24 DEBUG : Creating backend with remote "alist-aliyundrive:alist-backup/alist_backup_1_29_2023, 11_49_39 AM.json"
2023/02/12 13:34:24 DEBUG : Using config file from "/Users/gry/.config/rclone/rclone.conf"
2023/02/12 13:34:24 DEBUG : found headers: Referer,https://www.aliyundrive.com/drive,User-Agent,Mozilla/5.0 (Windows NT 6.1; Trident/7.0; rv:11.0) like Gecko
2023/02/12 13:34:25 DEBUG : fs cache: adding new entry for parent of "alist-aliyundrive:alist-backup/alist_backup_1_29_2023, 11_49_39 AM.json", "alist-aliyundrive:alist-backup"
2023/02/12 13:34:25 DEBUG : Creating backend with remote "."
2023/02/12 13:34:25 DEBUG : fs cache: renaming cache item "." to be canonical "/Users/gry"
2023/02/12 13:34:25 DEBUG : alist_backup_1_29_2023, 11_49_39 AM.json: Need to transfer - File not found at Destination
2023/02/12 13:34:26 INFO  : alist_backup_1_29_2023, 11_49_39 AM.json: Copied (new)
2023/02/12 13:34:26 INFO  :
Transferred:   	   10.474 KiB / 10.474 KiB, 100%, 0 B/s, ETA -
Transferred:            1 / 1, 100%
Elapsed time:         2.1s

2023/02/12 13:34:26 DEBUG : 8 go routines active
```

How can I bypass the `--header` option by editing the config?


Info:
```
$ rclone --version
rclone v1.61.1
- os/version: darwin 13.1 (64 bit)
- os/kernel: 22.2.0 (x86_64)
- os/type: darwin
- os/arch: amd64
- go/version: go1.19.4
- go/linking: dynamic
- go/tags: cmount
```

**Comment by simon-ding** [comment]
> > 2022/04/26 17:50:36 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
> > 2022/04/26 17:50:36 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> > 2022/04/26 17:50:36 DEBUG : HTTP REQUEST (req 0xc0008fdc00)
> > 2022/04/26 17:50:36 DEBUG : GET /**** HTTP/1.1
> > Host: bj29.cn-beijing.data.alicloudccp.com
> > User-Agent: rclone/v1.59.0-beta.6085.82e24f521.fix-6040-webdav
> > Depth: 0
> > Range: bytes=371982336-743916849
> > Referer: http://192.168.2.2:5244/dav/2019.mkv
> 
> I think that is the go standard library doing that - see [golang/go#44160](https://github.com/golang/go/issues/44160) so I don't think I can work around that at the moment so easily.
> 
> I've merged the fix above to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.59

hi. I don't think issue of golang will be fixed anytime sooner. Can we just provide a workaround? to handle the redirects ourself?

**Comment by ncw** [comment]
It looks like the Go team have merged a fix for this for go 1.21

This will land in rclone 1.64 I expect
